common/interpolation/include/interpolation/interpolation_utils.hpp

```diff
@@ -97,6 +144,38 @@ void validateKeysAndValues(
     throw std::invalid_argument("The size of base_keys and base_values are not the same.");
   }
 }
+
+/**
+ * Linear extrapolation
+ * */
+template <typename T, typename std::enable_if_t<std::is_floating_point_v<T>> * = nullptr>
+void lerp_extrapolate(
+  std::vector<T> const & base_keys, std::vector<T> const & base_values, T const & query_key,
+  T & query_value)
+{
+  if (query_key < base_keys[0]) {
+    auto const & t0 = base_keys[0];
+    auto const & t1 = base_keys[1];
+
+    auto const & y0 = base_values[0];
+    auto const & y1 = base_values[1];
+
+    auto const & ratio = (t0 - query_key) / (t1 - t0);
+    query_value = y0 - ratio * (y1 - y0);
+  }
+
+  if (query_key > base_keys.back()) {
+    auto const & tn = base_keys.rbegin()[0];
+    auto const & tn_1 = base_keys.rbegin()[1];
+
+    auto const & yn = base_values.rbegin()[0];
+    auto const & yn_1 = base_values.rbegin()[1];
+
+    auto const & ratio = (query_key - tn) / (tn - tn_1);
+    query_value = yn + ratio * (yn - yn_1);  // extrapolation and lerp have different semantics.
+  }
+}
```

Comment:
Is this implementation required?

https://github.com/autowarefoundation/autoware.universe/blob/a02830a142c6e8ad4b51c5174ba1a72d8237b829/common/interpolation/src/linear_interpolation.cpp#L21-L24
This lerp function allows extrapolation.

Reply:
Yes, it is the original implementation and is used within the code. The second lerp I have added has different arguments called by the nonlinear MPC.

Reply:
I mean that `lerp` already considers extrapolation, so I guess we do not have to add `lerp_extrapolate`.

What you wanna do in NMPC is linear interpolation allowing extrapolation with the following function, right?
```
std::vector<double> lerp(
  const std::vector<double> & base_keys, const std::vector<double> & base_values,
  const std::vector<double> & query_keys, bool const & extrapolate_end_points = false);
```
Then I'll add a change to your PR later.

Reply:
I haven't seen any extrapolation. In the MPC implementations, we sometimes add extra time or distance to guarantee MPC prediction length. At least the NMPC requires it.